File: src/hiloflo/modelos.py
"""Modelo de dominio de HILO-FLO (Mubea Florence, lineas ITW).

Vocabulario de planta:
  ITW      Induction Tempered Wire: las lineas de temple por induccion.
  Linea    ITW-1 .. ITW-14 (la 15 esta por instalarse). En SAP son los
           work centers BB001 .. BB014.
  Orden    un renglon del production schedule: N kg de un alambre de cierto
           diametro, ya asignado a un work center.
  Programa el conjunto de ordenes del horizonte que se esta programando.

La velocidad de linea viene en mm/s por diametro de alambre estirado
(WI-FLO-CSW-P-526), y de ahi sale el rendimiento en kg/h.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field, replace
from typing import Iterable

# --------------------------------------------------------------------------
# Equivalencia work center SAP <-> linea ITW
# --------------------------------------------------------------------------

def work_center_a_linea(work_center: str) -> str:
    """BB001 -> ITW-1. Si no reconoce el patron, regresa el texto tal cual."""
    m = re.fullmatch(r"\s*BB0*(\d+)\s*", str(work_center), re.I)
    return f"ITW-{int(m.group(1))}" if m else str(work_center).strip()


def linea_a_work_center(linea: str) -> str:
    """ITW-1 -> BB001."""
    m = re.fullmatch(r"\s*ITW[-\s]?(\d+)\s*", str(linea), re.I)
    return f"BB{int(m.group(1)):03d}" if m else str(linea).strip()


# --------------------------------------------------------------------------
# Catalogo de lineas
# --------------------------------------------------------------------------


@dataclass(frozen=True)
class Linea:
    """Una linea de temple por induccion."""

    clave: str  # "ITW-1"
    horas_disponibles: float = 0.0
    eficiencia: float = 1.0  # castiga la velocidad teorica de la receta
    minutos_cambio: float = 0.0  # cambio de medida / receta
    activa: bool = True  # ITW-15 sigue por instalarse

    def __post_init__(self) -> None:
        if not 0 < self.eficiencia <= 1:
            raise ValueError(f"{self.clave}: eficiencia debe estar en (0, 1]")
        if self.horas_disponibles < 0:
            raise ValueError(f"{self.clave}: horas_disponibles no puede ser negativa")

    @property
    def work_center(self) -> str:
        return linea_a_work_center(self.clave)


# --------------------------------------------------------------------------
# Velocidades de receta
# --------------------------------------------------------------------------


# Devanador que corre la linea cuando el schedule no indica otra cosa. El WI
# solo tabula el DEM para ITW-2, y su nota dice que el schedule lo marca en
# la seccion de notas; sin esa marca, corre el Neturen.
WINDER_PREDETERMINADO = "NETUREN"


@dataclass(frozen=True)
class PuntoVelocidad:
    """Una celda de la tabla ITW Line Speed del WI-FLO-CSW-P-526.

    Los campos ``winder``, ``grado`` y ``slm`` son los discriminantes que
    trae el documento: ITW-2 se tabula por devanador (Neturen / DEM) y por
    grado (9254 vs 1065); ITW-10 se tabula por SLM / NON SLM. ``None``
    significa "aplica a cualquiera".
    """

    linea: str
    diametro_mm: float
    mm_s: float
    winder: str | None = None  # "NETUREN" | "DEM"
    grado: str | None = None  # "9254" | "1065"
    slm: bool | None = None

    def __post_init__(self) -> None:
        if self.mm_s <= 0:
            raise ValueError(f"{self.linea}@{self.diametro_mm}: mm_s debe ser > 0")
        if self.diametro_mm <= 0:
            raise ValueError(f"{self.linea}: diametro_mm debe ser > 0")

    def aplica_a(self, orden: "Orden") -> bool:
        """True si esta receta es usable para la orden."""
        if self.winder is not None:
            # Sin devanador indicado en el schedule, solo aplica la receta del
            # devanador de planta; tomar la del DEM inflaria el rendimiento.
            esperado = orden.winder or WINDER_PREDETERMINADO
            if self.winder != esperado:
                return False
        if self.grado is not None and self.grado != orden.grupo_grado:
            return False
        if self.slm is not None and self.slm != orden.slm:
            return False
        return True

    @property
    def especificidad(self) -> int:
        """Cuantos discriminantes fija. Gana la receta mas especifica."""
        return sum(x is not None for x in (self.winder, self.grado, self.slm))


# --------------------------------------------------------------------------
# Ordenes y programa
# --------------------------------------------------------------------------


@dataclass(frozen=True)
class Orden:
    """Un renglon del production schedule."""

    id: str  # numero de orden SAP
    diametro_mm: float
    kilogramos: float
    linea: str  # "ITW-7"
    material: str = ""  # numero de material SAP
    descripcion: str = ""
    grupo_grado: str = "9254"  # "9254" (incluye 54SiCr6/60SiCr) o "1065"
    slm: bool = False
    winder: str | None = None  # "DEM" si el schedule lo indica en notas
    secuencia: int = 0
    notas: str = ""
    cliente_po: str = ""
    fijo: bool = False  # el optimizador no lo puede mover

    def __post_init__(self) -> None:
        if self.kilogramos <= 0:
            raise ValueError(f"orden {self.id}: kilogramos debe ser > 0")
        if self.diametro_mm <= 0:
            raise ValueError(f"orden {self.id}: diametro_mm debe ser > 0")

    @property
    def medida(self) -> str:
        """Etiqueta legible del diametro, para agrupar y reportar."""
        return f"{self.diametro_mm:.2f}"

    def mover_a(self, linea: str, secuencia: int | None = None) -> "Orden":
        return replace(
            self,
            linea=linea,
            secuencia=self.secuencia if secuencia is None else secuencia,
        )


@dataclass
class Programa:
    """El production schedule del horizonte."""

    ordenes: list[Orden] = field(default_factory=list)
    horizonte: str = ""

    def __iter__(self):
        return iter(self.ordenes)

    def __len__(self) -> int:
        return len(self.ordenes)

    @property
    def kilogramos(self) -> float:
        return sum(o.kilogramos for o in self.ordenes)

    def de_linea(self, linea: str) -> list[Orden]:
        """Ordenes de una linea, en la secuencia en que estan programadas."""
        return sorted(
            (o for o in self.ordenes if o.linea == linea),
            key=lambda o: (o.secuencia, o.id),
        )

    def lineas_usadas(self) -> list[str]:
        return sorted({o.linea for o in self.ordenes})

    def reemplazar(self, ordenes: Iterable[Orden]) -> "Programa":
        return Programa(ordenes=list(ordenes), horizonte=self.horizonte)

    def con_orden(self, orden: Orden) -> "Programa":
        """Copia del programa sustituyendo la orden con el mismo id."""
        return self.reemplazar(
            orden if o.id == orden.id else o for o in self.ordenes
        )

File: server/src/motor/modelos.js
/**
 * Modelo de dominio de HILO-FLO (Mubea Florence, lineas ITW).
 *
 * Vocabulario de planta:
 *   ITW      Induction Tempered Wire: las lineas de temple por induccion.
 *   Linea    ITW-1 .. ITW-14 (la 15 esta por instalarse). En SAP son los
 *            work centers BB001 .. BB014.
 *   Orden    un renglon del production schedule: N kg de un alambre de cierto
 *            diametro, ya asignado a un work center.
 *   Programa el conjunto de ordenes del horizonte que se esta programando.
 */

/**
 * Rango plausible de diametro de alambre estirado en las lineas ITW (mm).
 *
 * Lo usan los dos lectores: en el WI descarta renglones de encabezado que
 * traen numeros en la columna A, y en el schedule evita confundir el rango
 * de resistencia ("1950-2000 MPa") con un diametro.
 */
export const DIAMETRO_MIN = 4;
export const DIAMETRO_MAX = 30;

/**
 * Devanador que se asume cuando el schedule no indica cual se uso.
 *
 * El deber ser en Florence es el DEM, confirmado con planta, asi que es lo
 * que se calcula por omision. El WI dice que el schedule marca el DEM en la
 * seccion de notas, pero en la practica casi nunca lo anota: tomar esa
 * ausencia como "corrio con Neturen" subestimaria el rendimiento de ITW-2
 * hasta a la mitad (2.18x de diferencia a 5.72 mm).
 *
 * Que no venga anotado no se ignora: el analisis levanta un aviso con las
 * ordenes afectadas y con lo que costaria el cierre del programa si de
 * verdad hubieran corrido con Neturen. Ver servicio/analisis.js.
 */
export const WINDER_PREDETERMINADO = 'DEM';

/** El otro devanador. El schedule lo puede indicar explicitamente para
 *  registrar la excepcion al deber ser. */
export const WINDER_ALTERNO = 'NETUREN';

/** BB001 -> ITW-1. Si no reconoce el patron, regresa el texto tal cual. */
export function workCenterALinea(workCenter) {
  const m = /^\s*BB0*(\d+)\s*$/i.exec(String(workCenter ?? ''));
  return m ? `ITW-${Number(m[1])}` : String(workCenter ?? '').trim();
}

/** ITW-1 -> BB001. */
export function lineaAWorkCenter(linea) {
  const m = /^\s*ITW[-\s]?(\d+)\s*$/i.exec(String(linea ?? ''));
  return m ? `BB${String(Number(m[1])).padStart(3, '0')}` : String(linea ?? '').trim();
}

/** Ordena ITW-2 antes que ITW-10 (numerico, no alfabetico). */
export function ordenNatural(linea) {
  const m = /(\d+)/.exec(linea ?? '');
  return m ? Number(m[1]) : Number.MAX_SAFE_INTEGER;
}

export function compararLineas(a, b) {
  return ordenNatural(a) - ordenNatural(b) || String(a).localeCompare(String(b));
}

/** Una linea de temple por induccion. */
export class Linea {
  constructor({
    clave,
    horasDisponibles = 0,
    eficiencia = 1,
    minutosCambio = 0,
    activa = true,
  }) {
    if (!(eficiencia > 0 && eficiencia <= 1)) {
      throw new Error(`${clave}: eficiencia debe estar en (0, 1]`);
    }
    if (horasDisponibles < 0) {
      throw new Error(`${clave}: horasDisponibles no puede ser negativa`);
    }
    this.clave = clave;
    this.horasDisponibles = horasDisponibles;
    this.eficiencia = eficiencia;
    this.minutosCambio = minutosCambio;
    this.activa = activa;
    Object.freeze(this);
  }

  get workCenter() {
    return lineaAWorkCenter(this.clave);
  }
}

/**
 * Una celda de la tabla ITW Line Speed del WI-FLO-CSW-P-526.
 *
 * `winder`, `grado` y `slm` son los discriminantes que trae el documento:
 * ITW-2 se tabula por devanador (Neturen / DEM) y por grado (9254 vs 1065);
 * ITW-10 se tabula por SLM / NON SLM. `null` significa "aplica a cualquiera".
 */
export class PuntoVelocidad {
  constructor({ linea, diametroMm, mmS, winder = null, grado = null, slm = null }) {
    if (!(mmS > 0)) throw new Error(`${linea}@${diametroMm}: mmS debe ser > 0`);
    if (!(diametroMm > 0)) throw new Error(`${linea}: diametroMm debe ser > 0`);
    this.linea = linea;
    this.diametroMm = diametroMm;
    this.mmS = mmS;
    this.winder = winder;
    this.grado = grado;
    this.slm = slm;
    Object.freeze(this);
  }

  /** True si esta receta es usable para la orden. */
  /**
   * True si esta receta es usable para la orden.
   *
   * @param orden
   * @param winderEsperado  devanador que se exige; null acepta cualquiera.
   *   Lo decide TablaVelocidades, no la receta: primero se busca con el
   *   deber ser y solo si ese diametro no esta tabulado se reintenta
   *   abierto, para no perder una orden por un hueco del documento.
   */
  aplicaA(orden, winderEsperado = undefined) {
    const esperado =
      winderEsperado === undefined ? (orden.winder ?? WINDER_PREDETERMINADO) : winderEsperado;
    if (this.winder !== null && esperado !== null && this.winder !== esperado) return false;
    if (this.grado !== null && this.grado !== orden.grupoGrado) return false;
    if (this.slm !== null && this.slm !== orden.slm) return false;
    return true;
  }


  /** Cuantos discriminantes fija. Gana la receta mas especifica. */
  get especificidad() {
    return [this.winder, this.grado, this.slm].filter((x) => x !== null).length;
  }
}

/** Un renglon del production schedule. */
export class Orden {
  constructor({
    id,
    diametroMm,
    kilogramos,
    linea,
    material = '',
    descripcion = '',
    grupoGrado = '9254',
    slm = false,
    winder = null,
    secuencia = 0,
    notas = '',
    clientePo = '',
    fijo = false,
  }) {
    if (!(kilogramos > 0)) throw new Error(`orden ${id}: kilogramos debe ser > 0`);
    if (!(diametroMm > 0)) throw new Error(`orden ${id}: diametroMm debe ser > 0`);
    this.id = id;
    this.diametroMm = diametroMm;
    this.kilogramos = kilogramos;
    this.linea = linea;
    this.material = material;
    this.descripcion = descripcion;
    this.grupoGrado = grupoGrado;
    this.slm = slm;
    this.winder = winder;
    this.secuencia = secuencia;
    this.notas = notas;
    this.clientePo = clientePo;
    this.fijo = fijo;
    Object.freeze(this);
  }

  /** Etiqueta legible del diametro, para agrupar y reportar. */
  get medida() {
    return this.diametroMm.toFixed(2);
  }

  /** Lo unico de la orden que cambia la receta y el rendimiento. */
  get firma() {
    return `${this.diametroMm}|${this.grupoGrado}|${this.slm}|${this.winder}`;
  }

  moverA(linea, secuencia = null) {
    return new Orden({
      ...this,
      linea,
      secuencia: secuencia === null ? this.secuencia : secuencia,
    });
  }
}

/** El production schedule del horizonte. */
export class Programa {
  constructor(ordenes = [], horizonte = '') {
    this.ordenes = ordenes;
    this.horizonte = horizonte;
  }

  get length() {
    return this.ordenes.length;
  }

  get kilogramos() {
    return this.ordenes.reduce((t, o) => t + o.kilogramos, 0);
  }

  /** Ordenes de una linea, en la secuencia en que estan programadas. */
  deLinea(linea) {
    return this.ordenes
      .filter((o) => o.linea === linea)
      .sort((a, b) => a.secuencia - b.secuencia || String(a.id).localeCompare(String(b.id)));
  }

  lineasUsadas() {
    return [...new Set(this.ordenes.map((o) => o.linea))].sort(compararLineas);
  }

  reemplazar(ordenes) {
    return new Programa([...ordenes], this.horizonte);
  }
}
